R/methods-group-features.R

```diff
@@ -0,0 +1,1049 @@
+#' @param ... passed to the `groupFeatures` function on numeric values.
+#'
+#' @return input `XCMSnExp` with feature groups added (i.e. in column
+#'     `"feature_group"` of its `featureDefinitions` data frame.
+#'
+#' @family feature grouping methods
+#'
+#' @rdname groupFeatures-similar-rtime
+#'
+#' @importClassesFrom ProtGenerics Param
+#'
+#' @importFrom MsFeatures SimilarRtimeParam AbundanceSimilarityParam
+#'
+#' @importClassesFrom MsFeatures SimilarRtimeParam AbundanceSimilarityParam
+#'
+#' @importMethodsFrom MsFeatures groupFeatures featureGroups featureGroups<-
+#'
+#' @md
+#'
+#' @author Johannes Rainer
+#'
+#' @examples
+#'
+#' library(MsFeatures)
+#' ## Load a test data set with detected peaks
+#' data(faahko_sub)
+#' ## Update the path to the files for the local system
+#' dirname(faahko_sub) <- system.file("cdf/KO", package = "faahKO")
+#'
+#' ## Disable parallel processing for this example
+#' register(SerialParam())
+#'
+#' ## Group chromatographic peaks across samples
+#' xodg <- groupChromPeaks(faahko_sub, param = PeakDensityParam(sampleGroups = rep(1, 3)))
+#'
+#' ## Group features based on similar retention time (i.e. difference <= 2 seconds)
+#' xodg_grp <- groupFeatures(xodg, param = SimilarRtimeParam(diffRt = 2))
+#'
+#' ## Feature grouping get added to the featureDefinitions in column "feature_group"
+#' head(featureDefinitions(xodg_grp)$feature_group)
+#'
+#' table(featureDefinitions(xodg_grp)$feature_group)
+#' length(unique(featureDefinitions(xodg_grp)$feature_group))
+#'
+#' ## Using an alternative groupiing method that creates larger groups
+#' xodg_grp <- groupFeatures(xodg,
+#'     param = SimilarRtimeParam(diffRt = 2, groupFun = MsCoreUtils::group))
+#'
+#' length(unique(featureDefinitions(xodg_grp)$feature_group))
+NULL
+
+#' @rdname groupFeatures-similar-rtime
+#'
+#' @importMethodsFrom xcms hasFeatures featureDefinitions featureDefinitions<-
+#'
+#' @importFrom MsCoreUtils group
+#'
+#' @exportMethod groupFeatures
+#'
+#' @importClassesFrom xcms XCMSnExp XProcessHistory
+setMethod(
+    "groupFeatures",
+    signature(object = "XCMSnExp", param = "SimilarRtimeParam"),
+    function(object, param, msLevel = 1L, ...) {
+        fgs <- featureGroups(object)
+        if (length(msLevel) > 1)
+            stop("Currently only grouping of features from a single MS level",
+                 " is supported.", call. = FALSE)
+        if (any(colnames(featureDefinitions(object)) == "ms_level"))
+            is_msLevel <- featureDefinitions(object)$ms_level == msLevel
+        else is_msLevel <- rep(TRUE, nrow(featureDefinitions(object)))
+        if (all(is.na(fgs)))
+            fgs <- rep("FG", length(fgs))
+        fgs[!is_msLevel] <- NA
+        nas <- is.na(fgs)
+        fgs <- factor(fgs, levels = unique(fgs))
+        rtl <- split(featureDefinitions(object)$rtmed, fgs)
+        res <- lapply(
+            rtl, function(z, param)
+                MsFeatures:::.format_id(groupFeatures(z, param = param, ...)),
```

Comment:
Should that eventually be exported by MsFeatures ?

Reply:
yes/no. For now it's a private function there, but it's a valid point. I will consider/rethink that.